### PR TITLE
alternative, much simpler algorithm for strict func checking

### PR DESCRIPTION
**Pull request by Araq**
Todo: 

- [x] Document it.
- [x] Fix the regressions.

### CODE REVIEW

**Comment by Araq** [comment]
Ping @ee7, feedback appreciated.

**Comment by ShalokShalom** [comment]
How is the coverage?

Does the old way cover just as many cases?

How strict are strict funcs now?
I consider the design decision, to leave the strictness on libraries in C++ etc up to the author of the binding, slightly irritating. 

Can we make it default, that external libraries in other languages are strict by default?

**Comment by Araq** [comment]
You seem to interpret "strict func" in some sense that has very little to do with its (old or new) definition and this is not the right place to answer these questions.

**Comment by ee7** [comment]
Looks great at first glance. I haven't reviewed `semstrictfuncs.nim` deeply, but resolving the outstanding `strictFuncs` issues without introducing regressions would be wonderful. Seems like a big step towards making strict funcs the default.

I've checked that the latest commit (b6d219e1ec5b64725481eca6a12c282e7ab175c0) in this PR also produces the expected behavior for snippets in #20808. Can we add the below test case, and make this PR close that issue too?

```Nim
{.experimental: "strictFuncs".}

type
  Node = ref object
    num: int

func mutate(nodes: seq[Node]) =
  for node in nodes:
    node.num = 42

let nodes = @[Node(num: 7)]
mutate(nodes)
```

with the expected output of something like:

```
/tmp/foo.nim(9, 9) Error: cannot mutate location node.num within a strict func
```

---

> Make it opt-out for v2!?

Regarding Nim 2.0 (or 2.2) and strict funcs: what do we think about making strict funcs the default, but producing a warning for a transition period rather than an error?

I think our options for Nim 2.0 are:

1. Keep the status quo: strict funcs are opt-in via `--experimental:strictFuncs`. However, the compiler may now produce an error for code that previously compiled with `strictFuncs`, but mutated in ways that were incorrect by the spec.
2. Make strict funcs the default, but make it produce a warning. Allow making it an error via whatever mechanism is preferred (e.g. `--strictFuncs:error`, or `--warningAsError:strictFuncs`). This is non-breaking: code that previously compiled with or without `strictFuncs` continues to compile.
3. Make strict funcs the default, and make it produce an error. This breaks code that either:
   - did not use `strictFuncs`, but does mutate in ways that `strictFuncs` would have complained about
   - or did use `strictFuncs`, but mutated in ways that `strictFuncs` did not catch before this PR

where option 3 seems too aggressive.

File: doc/manual_experimental.md [hunk 1]
```diff
@@ -519,8 +519,7 @@ Since version 1.4, a stricter definition of "side effect" is available.
 In addition to the existing rule that a side effect is calling a function
 with side effects, the following rule is also enforced:
 
-Any mutation to an object does count as a side effect if that object is reachable
-via a parameter that is not declared as a `var` parameter.
+A store to the heap via a `ref` or `ptr` indirection is not allowed.
```
Comment (ee7) [inline doc/manual_experimental.md:522]: So it's intended that the below code will no longer compile?

```Nim
{.experimental: "strictFuncs".}

type
  Node = ref object
    num: int

func mutate(node: var Node) =
  node.num = 42

var node = Node(num: 7)
mutate(node)
```

With b6d219e1ec5b64725481eca6a12c282e7ab175c0, Nim produces an error:

```
/tmp/foo.nim(8, 7) Error: cannot mutate location node.num within a strict func
```

Comment (Araq) [reply doc/manual_experimental.md:522]: Yes, it is.

Comment (ee7) [reply doc/manual_experimental.md:522]: Ah. Then I think that I prefer the previous rule for strictFuncs. I liked being able to write the above.

Could you expand on the rationale for the rule change? The new rule is described as "much simpler and hopefully easier to teach than the old but more advanced algorithm" - was teaching the old rule a problem? Can we resolve the open `strictFuncs` issues without changing the rule? (To others: see https://github.com/nim-lang/RFCs/issues/234 for background on the rule).

I think that removing the ability for a strict func to mutate a `ref` parameter marked as `var` decreases the usefulness of strict funcs significantly. For a procedure that does nothing but mutate a `ref object` parameter, I'd prefer to read

```Nim
func foo(node: var Node) =
```

than

```Nim
proc foo(node: var Node) =
```

or

```Nim
proc foo(node: Node) =
```

Comment (snej) [reply doc/manual_experimental.md:522]: The new definition (if I understand it correctly) is much more in line with the meaning of “function” in mathematics and functional programming, and of “pure function” in some non-functional languages. (For example it matches the meaning of the `__attribute__((pure))` annotation in GCC and Clang.)

**Comment by Araq** [comment]
> I'd prefer to read
> func foo(node: var Node) =

That creates a pointer to a pointer and it was never my intention to encourage this style. It also doesn't work with inheritance as the double indirection disables the subtype relation.

**Comment by alaviss** [comment]
This is incompatible with custom exceptions:

```nim
func newPrematureCloseDefect*(id: int): ref PrematureCloseDefect =
  ## Creates a `PrematureCloseDefect`
  result = newException(PrematureCloseDefect):
    "Resource id " & $id & " was invalidated before its unregistered"
  result.id = id
```

I can workaround this by using object construction, but this will inevitably break a lot of code if it ever become the default.

**Comment by Araq** [comment]
You really should use object construction either way.

**Pull request merged**